### PR TITLE
Document generated columns and proctime()

### LINKED ISSUE #759: Document proctime()
### Related code PR

https://github.com/risingwavelabs/risingwave/pull/9088

This PR is part of the new [generated columns](https://github.com/risingwavelabs/risingwave/issues/8699).

### Which part(s) of the docs might be affected or should be updated? And how?

- Add a new function proctime() in [SQL references - Functions and operators - Date/time](https://www.risingwave.dev/docs/current/sql-function-datetime/) 

- Document the proctime() attribute in [CREATE TABLE](https://www.risingwave.dev/docs/current/sql-create-table/) & [CREATE SOURCE](https://www.risingwave.dev/docs/current/sql-create-source/)

### Reference

https://nightlies.apache.org/flink/flink-docs-master/docs/dev/table/concepts/time_attributes/#processing-time

Introduction (generated by AI, edited):

PROCTIME() allows users to specify a processing time on SOURCE/TABLE. The function returns the current system time when a record is processed, which can be useful for several purposes, including auditing and tracking data changes over time.

To use PROCTIME(), you can define a new table with a column of type `timestamptz` and set its default value to PROCTIME(). For example:

`CREATE TABLE t1 (proc_time Timestamptz as PROCTIME());`

This creates a new table named t1 with a column named `proc_time` of type `timestamptz`. Whenever a new record is inserted into the table, the value of `proc_time` is automatically set to the current processing time using the PROCTIME() function.

### LINKED ISSUE #761: Document generated columns
### Related code PR

main: https://github.com/risingwavelabs/risingwave/pull/8700
supporting: https://github.com/risingwavelabs/risingwave/pull/8841
supporting: https://github.com/risingwavelabs/risingwave/pull/8779

These PRs are part of [generated columns](https://github.com/risingwavelabs/risingwave/issues/8699).

### Which part(s) of the docs might be affected or should be updated? And how?

Document the generated columns in [CREATE TABLE](https://www.risingwave.dev/docs/current/sql-create-table/) & [CREATE SOURCE](https://www.risingwave.dev/docs/current/sql-create-source/)

### Reference

Read the Documentation section at https://github.com/risingwavelabs/risingwave/pull/8700.

SQL example:
https://github.com/risingwavelabs/risingwave/pull/8700/files#diff-e6725651ecb87465c410915b51ccab47573e8dc81c134a25d3548b6fcf42f3cd
(Note that the syntax is closer to Flink, not standard PostgreSQL)

PostgreSQL: https://www.postgresql.org/docs/current/ddl-generated-columns.html
Other popular RDBMSes support generated columns as “computed columns” or “virtual columns.”

### PR DESCRIPTION
**Pull request by hengm3467**
## Info
- **Description**: 
Document generated columns and proctime() function.

- Add a new article "Generated columns" under "Query syntax"
- Update `CREATE TABLE` and `CREATE SOURCE` to include descriptions of generated columns and proctime().
- Update "Date and time functions" to include `proctime()`.

- **Preview**: 
[ Paste the preview link to the edited page here. ]

- **Related code PR**: 
[ Provide the link to the code PR here. For example, https://github.com/risingwavelabs/risingwave/pull/4085. Delete this part if there's no code PR related. ]

- **Related doc issue**: 
Resolves [ Provide the link to the doc issue here. ]

- **Notes**: 
[ Any additional information? ]

## Before merging
  - [ ] (For version-specific PR) I have selected the corresponding software version in **Milestone** and linked the related doc issue to this PR in **Development**.
  - [ ] I have acquired the approval from the owner (and optionally the reviewers) of the code PR and at least one tech writer (`bernscode`, `CharlieSYH`, `emile-00`, & `hengm3467`). 
  - [ ] I have checked the doc site preview, and the updated parts look good. <details><summary>How?</summary>Scroll down and open this link: image
</details>

### CODE REVIEW

File: docs/sql/query-syntax/query-syntax-generated-columns.md [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+
+```sql
+CREATE TABLE t1 (v1 int AS v2-1, v2 int, v3 int AS v2+1);
+```
+
+Note that a generation expression cannot reference another generated column.
+
+A generated column in a table is slightly different from one in a source.
+
+A generated column in a table is stored in the created table, and computed when it is inserted or updated. This is equivalent to the `STORED` type of generated columns in PostgreSQL.
```
Comment (CharlieSYH) [inline docs/sql/query-syntax/query-syntax-generated-columns.md]: Consider adding bullet points in front of the two comparison sentences.

**Review by CharlieSYH — APPROVED** [review]
LGTM.
Btw, what's the markdown formatter you're using? Seems pretty good.

File: docs/sql/commands/sql-create-source.md [hunk 2]
```diff
@@ -24,14 +24,20 @@ ROW FORMAT data_format
 [ ROW SCHEMA LOCATION [ 'location' | CONFLUENT SCHEMA REGISTRY 'schema_registry_url' ] ];
 ```
 
+For the syntax of `schema_definition`, see [Parameters](sql-create-table.md#parameters) in `CREATE TABLE`.
+
+:::note
+To know when a data record is loaded to RisingWave, you can define a column that is generated based on the processing time (`<column_name> AS proctime()`) when creating the table or source.
```
Comment (xxchan) [inline docs/sql/commands/sql-create-source.md]:
```suggestion
To know when a data record is loaded to RisingWave, you can define a column that is generated based on the processing time (`<column_name> timestampz AS proctime()`) when creating the table or source.
```

File: docs/sql/commands/sql-create-table.md [hunk 3]
```diff
@@ -114,29 +108,22 @@ export const svgTwo = rr.Diagram(
...
 
-:::note
 Names and unquoted identifiers are case-insensitive. Therefore, you must double-quote any of these fields for them to be case-sensitive.
-:::
 
-:::note
 If creating a materialized source, remember to include the connector settings with the `WITH` clause and to specify the data format with the `ROW FORMAT` clause. See [`CREATE SOURCE`](sql-create-source.md) for a full list of supported connectors and data formats.
-:::
+
+To know when a data record is loaded to RisingWave, you can define a column that is generated based on the processing time (`<column_name> AS proctime()`) when creating the table or source.
```
Comment (xxchan) [inline docs/sql/commands/sql-create-table.md]:
```suggestion
To know when a data record is loaded to RisingWave, you can define a column that is generated based on the processing time (`<column_name> timestampz AS proctime()`) when creating the table or source.
```

File: docs/sql/query-syntax/query-syntax-generated-columns.md (hunk 1 above)
Comment (yuhao-su) [reply docs/sql/query-syntax/query-syntax-generated-columns.md]: > computed when it is inserted or updated

Currently we do not allow delete/update a table if it contains any generated column. We may also not users about this

File: docs/sql/functions-operators/sql-function-datetime.md [hunk 4]
```diff
@@ -50,15 +44,14 @@ title: Date and time functions and operators
...
 | extract ( *field* FROM *source* ) → *numeric_output* | Extracts the value of a date or timestamp. <br /> *field* can be year, month, day, hour, minute, second, doy, or dow. <br /> *source* can be the date or timestamp.|`extract(day from date '2022-04-07')` → `7` <br /> `extract(hour from timestamp '2022-04-07 22:00:30')` → `22`|
 | extract ( epoch FROM *timestamp_with_time_zone* ) → *seconds_numeric_output* | Converts the value of timestamp with time zone to Unix epoch seconds (the number of seconds since 1970-01-01 00:00:00 UTC). Negative for timestamps prior to that.|`extract(epoch from '2010-01-01 12:34:56.789012Z'::timestamp with time zone)` → `1262349296.789012`|
+| proctime() | Returns the system time when a record is processed. You can use this function to specify the processing time of a record in a table or source. | You can define a table with a generated column of type `timestamptz` and set its value to `proctime()`. For example: `CREATE TABLE t1 (v1 int, proc_time Timestamptz as PROCTIME()) ;`|
```
Comment (xxchan) [inline docs/sql/functions-operators/sql-function-datetime.md]:
```suggestion
| proctime() → *timestamp_with_time_zone* | Returns the system time when a record is processed. You can use this function to specify the processing time of a record in a table or source. | You can define a table with a generated column of type `timestamptz` and set its value to `proctime()`. For example: `CREATE TABLE t1 (v1 int, proc_time timestamptz as PROCTIME()) ;`|
```

Not sure what's the desired format. Just a suggestion.

Comment (xxchan) [reply docs/sql/functions-operators/sql-function-datetime.md]: looks inconsistent 😇
image

Comment (yuhao-su) [inline docs/sql/functions-operators/sql-function-datetime.md]: > You can define a table with a generated column of type `timestamptz` and set its value to `proctime()`

We can actually define any return time that can be cast from `Timestamptz`. The return type of the `proctime` function is `Timestamptz`.

Comment (hengm3467) [reply docs/sql/functions-operators/sql-function-datetime.md]: Just removed this redundant statement.

**Comment by hengm3467** [comment]
@yuhao-su & @xxchan, I updated the topics based on your comments. Please take a look again. Thanks! :)

**Comment by yuhao-su** [comment]
The generated column part LGTM

**Review by xxchan — APPROVED** [review]
generally LGTM

**Pull request merged**